From bbf65740f07829e986bc89c21b373e04fa342f78 Mon Sep 17 00:00:00 2001
From: Pratham Patel <ppatel@ciq.com>
Date: Fri, 8 Aug 2025 14:39:45 +0530
Subject: [PATCH] dm ioctl: prevent potential spectre v1 gadget

jira VULN-49983
cve CVE-2022-49122
commit-author Jordy Zomer <jordy@jordyzomer.github.io>
commit cd9c88da171a62c4b0f1c70e50c75845969fbc18

It appears like cmd could be a Spectre v1 gadget as it's supplied by a
user and used as an array index. Prevent the contents of kernel memory
from being leaked to userspace via speculative execution by using
array_index_nospec.

	Signed-off-by: Jordy Zomer <jordy@pwning.systems>
	Signed-off-by: Mike Snitzer <snitzer@redhat.com>
(cherry picked from commit cd9c88da171a62c4b0f1c70e50c75845969fbc18)
	Signed-off-by: Pratham Patel <ppatel@ciq.com>
---
 drivers/md/dm-ioctl.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/drivers/md/dm-ioctl.c b/drivers/md/dm-ioctl.c
index 00ee0e0b399c..8b2896c62205 100644
--- a/drivers/md/dm-ioctl.c
+++ b/drivers/md/dm-ioctl.c
@@ -18,6 +18,7 @@
 #include <linux/dm-ioctl.h>
 #include <linux/hdreg.h>
 #include <linux/compat.h>
+#include <linux/nospec.h>
 
 #include <linux/uaccess.h>
 
@@ -1731,6 +1732,7 @@ static ioctl_fn lookup_ioctl(unsigned int cmd, int *ioctl_flags)
 	if (unlikely(cmd >= ARRAY_SIZE(_ioctls)))
 		return NULL;
 
+	cmd = array_index_nospec(cmd, ARRAY_SIZE(_ioctls));
 	*ioctl_flags = _ioctls[cmd].flags;
 	return _ioctls[cmd].fn;
 }
-- 
2.47.3

